File: backend/entities/ProjectStatusTransition.cs
using System;

namespace Pims.Dal.Entities
{
    /// <summary>
    /// ProjectStatusTransition class, provides a way to manage workflow project status transitions within the solution.
    /// </summary>
    public class ProjectStatusTransition : BaseEntity
    {
        #region Properties
        /// <summary>
        /// get/set - Foreign key to the workflow project status this transitions is from.
        /// </summary>
        public int FromWorkflowId { get; set; }

        /// <summary>
        /// get/set - Foreign key to the workflow project status this transition is from.
        /// </summary>
        public int FromStatusId { get; set; }

        /// <summary>
        /// get/set - The workflow project status this transition is from.
        /// </summary>
        public WorkflowProjectStatus FromWorkflowStatus { get; set; }

        /// <summary>
        /// get/set - The action name that describes the transition.
        /// </summary>
        public string Action { get; set; }

        /// <summary>
        /// get/set - Foreign key to the workflow project status this goes to.
        /// </summary>
        public int ToWorkflowId { get; set; }

        /// <summary>
        /// get/set - Foreign key to the workflow project status this goes to.
        /// </summary>
        public int ToStatusId { get; set; }

        /// <summary>
        /// get/set - Whether this status requires validating prior tasks before transitioning.
        /// </summary>
        public bool ValidateTasks { get; set; } = true;

        /// <summary>
        /// get/set - The workflow project status this goes to.
        /// </summary>
        public WorkflowProjectStatus ToWorkflowStatus { get; set; }
        #endregion

        #region Constructors
        /// <summary>
        /// Creates a new instance of a ProjectStatusTransition object.
        /// </summary>
        public ProjectStatusTransition() { }

        /// <summary>
        /// Creates a new instance of a ProjectStatusTransition object, initializes it with specified arguments.
        /// </summary>
        /// <param name="from"></param>
        /// <param name="action"></param>
        /// <param name="to"></param>
        public ProjectStatusTransition(WorkflowProjectStatus from, string action, WorkflowProjectStatus to)
        {
            this.FromWorkflowStatus = from ?? throw new ArgumentNullException(nameof(from));
            this.FromWorkflowId = from.WorkflowId;
            this.FromStatusId = from.StatusId;
            from.ToStatus.Add(this);
            this.Action = action;
            this.ToWorkflowStatus = to ?? throw new ArgumentNullException(nameof(to));
            this.ToWorkflowId = to.WorkflowId;
            this.ToStatusId = to.StatusId;
            to.FromStatus.Add(this);
        }
        #endregion
    }
}

File: express-api/src/typeorm/Entities/ProjectStatusTransition.ts
import { Entity, Column, Index, ManyToOne, JoinColumn, Relation, PrimaryColumn } from 'typeorm';
import { Workflow } from './Workflow';
import { ProjectStatus } from './ProjectStatus';
import { BaseEntity } from '@/typeorm/Entities/abstractEntities/BaseEntity';

@Entity()
@Index(['ToWorkflowId', 'ToStatusId'])
@Index(['FromWorkflowId', 'FromStatusId'])
export class ProjectStatusTransition extends BaseEntity {
  // From Workflow Relation
  @PrimaryColumn({ name: 'from_workflow_id', type: 'int' })
  FromWorkflowId: number;

  @ManyToOne(() => Workflow, (Workflow) => Workflow.Id)
  @JoinColumn({ name: 'from_workflow_id' })
  FromWorkflow: Relation<Workflow>;

  // From Status Relation
  @PrimaryColumn({ name: 'from_status_id', type: 'int' })
  FromStatusId: number;

  @ManyToOne(() => ProjectStatus, (ProjectStatus) => ProjectStatus.Id)
  @JoinColumn({ name: 'from_status_id' })
  FromStatus: Relation<ProjectStatus>;

  // To Workflow Relation
  @PrimaryColumn({ name: 'to_workflow_id', type: 'int' })
  ToWorkflowId: number;

  @ManyToOne(() => Workflow, (Workflow) => Workflow.Id)
  @JoinColumn({ name: 'to_workflow_id' })
  ToWorkflow: Relation<Workflow>;

  // To Status Relation
  @PrimaryColumn({ name: 'to_status_id', type: 'int' })
  ToStatusId: number;

  @ManyToOne(() => ProjectStatus, (ProjectStatus) => ProjectStatus.Id)
  @JoinColumn({ name: 'to_status_id' })
  ToStatus: Relation<ProjectStatus>;

  @Column({ type: 'character varying', length: 100 })
  Action: string;

  @Column('boolean')
  ValidateTasks: boolean;
}
